### DISCUSSION TITLE
ROADMAP for 2021

### DISCUSSION THREAD

**Discussion by remiadon** [scikit-mine/scikit-mine#115] [Q&A]
As mentionned by @chuanconggao in https://github.com/scikit-mine/scikit-mine/pull/109#issuecomment-820287161

we now share the project roadmap for transparency

So here is what I have in mind right now (this is prone to later modification)
- may 2021 : 
    - periodic pattern with an MDL criterion, [paper here](https://hal.archives-ouvertes.fr/hal-01951722/document)
    - scikit-learn compatible Transformer for SLIM
 - june 2021: 
    - DiffNorm [slides here](https://icfca2017.irisa.fr/files/2017/06/Vreeken_keynote_slides.pdf), living, in skmine.multiclass
    - SeqKrimp / GoKrimp ?? living in skmine.sequence
  - september 2021:
      - SQS, living in skmine.sequence, [paper here](https://arxiv.org/pdf/1602.05012.pdf) 

Needless to say, other algorithms are welcome, as long as they follow the [inclusion criteria](https://scikit-mine.github.io/scikit-mine/CONTRIBUTING.html#inclusion-criteria)

**Comment by chuanconggao** [comment]
Cool. Is there any suggested areas for long-term (in a few years) coverage, like graph algorithm, stream algorithm, etc.?

BTW, in long term, for many included algorithms, the implementations may share some common code, like common data structure or utility functions. It would help form better abstraction, code style consistency, and maintainability. Not sure whether it should be covered in roadmap (if dedicated effort is required) or contributing guideline.

**Reply by remiadon** [reply]
We have one graph algorithm to come (mining sub-graph based on MDL)
Stream mining is not yet considered, but again if the community asks for it, we will have to consider it seriously. My only concern is about the type of input, usually graphs are represented using scipy.sparse matrices.
From en engineering side I think we should provide compatibility with the rest of the Python ecosystem. For example, for stream mining [River ML](https://github.com/online-ml/river), a library for continuous ML, exposes methods similar to sklearn, but suffixed by `_one` (eg. `predict` becomes `predict_one`). For skmine I think a `fit_one` method would make sense if to implement stream mining algorithms.

--------------------------

If you ask me about "in a few years", skmine is meant to be included in scikit-learn's contrib, where other people already expressed their need for "associative learning" algorithms ([here](https://github.com/scikit-learn/scikit-learn/issues/2662) is the thread).

 What they mean by associative learning is mining association rules + taking decision based on those rules. Although this was not our first idea when starting skmine, maybe we should discuss this part. IMO mining rules is a different topic, and should not be included, but some may disagree.

--------------------------

Concerning common code / data structure etc.
All methods I implemented use :
 - numpy for array operations / inner function speedups (eg  in `.fit`)
 - pandas for high level manipulations / expositions (eg. in `.discover`)
 - roaring bitmaps to represent transaction ids (tids) and perform fast intersection / union over them.

numpy and pandas are used by the entire scientific comunity. The one data structure I want to introduce is the roaring bitmap, I find it suitable for mining tasks / optimizations. Without those bitmaps all methods in `skmine.itemsets` would take ages to run ...
But maybe this needs to be clarified on a dedicated page in the doc ?

For utility functions, you can already find `utils.py` files in all skmine sub modules, but there are subject to evolution of course

Hope it answers some of your questions

**Reply by chuanconggao** [reply]
Personally I think rule mining is kinda of covered by pattern mining, with specific requirements/constraints. It makes sense to cover as well, especially when underlaying algorithms can be reused. 

Regarding reusing data structures, there may be more to consider in future, like FP-tree (many people may even directly use for their own algorithms). Agree we can consider later but may need to remind contributors to take into consideration when submitting PR.

**Reply by lgalarra** [reply]
I agree with @chuanconggao. Rules are like any other type of pattern. It turns out that you can build them trivially from simpler patterns (itemsets, sequences), but it does not mean it is the most efficient way to do it. And like patterns, you can present them to users or plug them into ML models that will use them as features.

**Reply by remiadon** [reply]
@lgalarra and @chuanconggao  I guess I was living in a simple world ;)

On the long term I Imagine the following
 - a RuleMiner class, for generic rule mining from pre-mined patterns
 - other classes that output rules, and not patterns, for specific purposes
 - reusable base datastructures (like the FP-tree)

would that be OK ?

@lgalarra concerning the usage of rules as features to ML models, I honestly don't know how to do that, but I guess we'll fugure out ;)

**Comment by lgalarra** [comment]
Hi Rémi,

Indeed the use of association rules for prediction (usually classification) is less common but it exists. I should make a clarification here. I am not talking about methods that learn classifiers based on decisions lists or sets (https://christophm.github.io/interpretable-ml-book/rules.html). Those methods learn the rules as part of the classifier induction process, which is different to learning the rules via pattern mining first and then plugging them into another classifier. I can think of types of methods:

- Use the rules directly as an ensemble assuming that the consequent of the rules defines the classes among which you want to discriminate. Take all the rules and aggregate the answers of those that apply for a given example (pretty much like a random forest). If there are ties or contradictions, i.e., the antecedent of more than one rule triggers, the answer of the rule with the highest confidence (then support, then whatever metric you like) prevails. http://www.philippe-fournier-viger.com/sequential_rules_prediction_2012.pdf, https://dl.acm.org/doi/pdf/10.1145/3018661.3018739
- Learn a classifier using those rules as features. Your instances become vectors where each rule becomes a component. You put, e.g.,  a '1' if the rule predicts that instance, 0 otherwise. Then you build a matrix with your instances and learn the classifier you like, e.g., LR, SVM, etc. That way an ML reweighs the pertinence of the rules in the classification.  This is less common though, https://arxiv.org/pdf/1812.10384.pdf, https://etda.libraries.psu.edu/files/final_submissions/7404

Best,
Luis
